Yuhsuan/1277 fitting error

YuHsuan-Hwang opened the pull request:
**Description**

* What is implemented or fixed? Mention the linked issue(s), if any.

1. Fixed #1277 by implementing error calculations based on [Condon (1997)](https://ui.adsabs.harvard.edu/abs/1997PASP..109..166C/abstract), which is also used by [casa imfit](https://casadocs.readthedocs.io/en/latest/api/tt/casatasks.analysis.imfit.html).
    * When the beam table is empty, we assume the samples are independent and apply formulas in Condon (1997) section 2 eqn 21.
    * When the beam table is not empty, we assume there is correlated noise and apply formulas in Condon (1997) section 3 eqn 41 and 42.
    * If the beam table is not empty but the beam of the current plane is not found, we apply the first beam.
    * If the background offset is a free parameter during fitting, background offset error is calculated by the covariance matrix.

2. Changed from unweighted fitting to weighted fitting with the information of image standard deviation.
    * Please note that this will change the iteration number and the fitting results and thus the execution time.

3. Added calculation of integrated flux in the fitting result. This requires frontend branch https://github.com/CARTAvis/carta-frontend/pull/2229 and protobuf branch https://github.com/CARTAvis/carta-protobuf/pull/88.
    * Applied Condon (1997) section 2 eqn 21 and section 3 eqn 42 for the calculation.
    * Applicable only when image units are exactly "Jy/pixel" or "Jy/beam." The calculated flux unit is Jy.
    * When both amplitude and FWHM sizes of a component are fixed, the integrated flux becomes fixed.

* How does this PR solve the issue? Give a brief summary.

    * Obtained beam size, image unit, ~and image std (of the entire image or regions)~ in `Frame.FitImage`.
    * Obtained Median Absolute Deviation (MAD) as image std in `ImageFitter.SolveSystem`.
    * Applied weighted fitting (requires image std) in `ImageFitter.SolveSystem`.
    * Added error calculation (requires beam size and image std) and integrated flux calculation (requires image unit) in `ImageFitter.CalculateErrors`.

* Are there any companion PRs (frontend, protobuf)?

    * https://github.com/CARTAvis/carta-frontend/pull/2229
    * https://github.com/CARTAvis/carta-protobuf/pull/88

* Is there anything else that testers should know (e.g. exactly how to reproduce the issue)?

* ~ToDo~
- [x] ~code maintenance: add documentation in `ImageFitter.h`.~

**Checklist**

- [x] changelog updated / ~no changelog update needed~
- [x] ~e2e test passing~ / added corresponding fix
- [ ] protobuf updated to the latest dev commit / ~no protobuf update needed~
- [x] added reviewers and assignee
- [x] added ZenHub estimate, milestone, and release

YuHsuan-Hwang committed:
get image std

YuHsuan-Hwang committed:
apply weighted fitting

YuHsuan-Hwang committed:
get beam size

YuHsuan-Hwang committed:
fix unit test

YuHsuan-Hwang committed:
change error calculation

YuHsuan-Hwang committed:
calculate integrated flux; fix background error

YuHsuan-Hwang committed:
update protobuf

confluence commented on src/Frame/Frame.cc:
```diff
@@ -1773,17 +1803,32 @@ bool Frame::FitImage(const CARTA::FittingRequest& fitting_request, CARTA::Fittin
+            float image_std;
+            std::vector<CARTA::StatsType> required_stats = {CARTA::StatsType::Sigma};
+            bool per_z = false;
+            std::map<CARTA::StatsType, std::vector<double>> stats_map;
```
This should probably be `unordered_map`, but I see that it's `map` in the existing code. There are a few other places where we're using `map`; this should probably be raised in a separate PR (I don't think there are any cases where we actually need an ordered map -- ensuring that we use `unordered_map` may improve performance, but `map` uses less memory, so perhaps this warrants further discussion).

YuHsuan-Hwang replied:
@confluence Yes, `GetRegionStats` requires `map`. Looks like we don't set up a long list of stats type, so it's fine to use `map` for now.
Do you prefer I add a TODO in `GetRegionStats` or file a code maintenance issue?

confluence replied:
I think a more general code maintenance issue would be good.

YuHsuan-Hwang replied:
Filed #1295.

YuHsuan-Hwang committed:
add doc

YuHsuan-Hwang committed:
change std calculation

YuHsuan-Hwang committed:
add doc; reformat

YuHsuan-Hwang committed:
fix center error

YuHsuan-Hwang commented:
New changes:
* Added documentation in `ImageFitter.h`.
* Replaced std of the image / region with Median Absolute Deviation (MAD) calculation for weighted fitting and error calculation.
* Fixed incorrect center error calculation.

YuHsuan-Hwang committed:
update protobuf

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1277_fitting_error

YuHsuan-Hwang committed:
update changelog

kswang1029 commented:
this is looking good. I have adjusted e2e tests accordingly.

kswang1029 commented on src/ImageFitter/ImageFitter.cc:
```diff
@@ -223,26 +240,112 @@ int ImageFitter::SolveSystem(CARTA::FittingSolverType solver) {
+            pa_err = sqrt(4.0 * tmp * tmp / rho_square_3) * 180.0 / M_PI;
+
+            if (_unit == "Jy/beam") {
+                const double flux = 2 * M_PI * fwhm_x * fwhm_y * SQ_FWHM_TO_SIGMA * amp / _beam_size / _beam_size;
```
need to divide a solid angle conversion factor as pi / 4.0 / ln(2.0) ~ 1.1330900354567985

`flux = 2 * M_PI * fwhm_x * fwhm_y * SQ_FWHM_TO_SIGMA * amp / _beam_size / _beam_size / (M_PI / 4.0 / log(2.0))`

YuHsuan-Hwang committed:
fix flux calculation

confluence committed:
Updated protobuf version.
